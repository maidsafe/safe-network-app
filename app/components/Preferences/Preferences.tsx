import React, { Component } from 'react';
import List from '@material-ui/core/List';

import { PreferenceItem } from './PreferenceItem';

import { UserPreferences } from '$Definitions/application.d';
import { generateRandomString } from '$Utils/app_utils';
import {
    launchOnLogin,
    pinLaunchpadToMenu,
    releaseLaunchpadFromMenu,
    storeUserPreferences
} from '$App/actions/launchpad_actions';

interface Props {
    userPreferences: UserPreferences;
    onChange: Function;
    requiredItems?: { [item: string]: boolean };
}

export class Preferences extends Component<Props> {
    static defaultProps = {
        userPreference: {},
        requiredItems: {
            autoUpdate: true,
            pinToMenuBar: true,
            launchOnStart: true,
            showDeveloperApps: true,
            warnOnAccessingClearnet: true
        },
        onChange: () => {}
    };

    public static changeCompleted( userPreferences ) {
        // Enable or disable auto launch
        launchOnLogin( userPreferences.launchOnStart );

        // switch between standard or tray window
        if ( userPreferences.pinToMenuBar ) {
            pinLaunchpadToMenu();
        } else {
            releaseLaunchpadFromMenu();
        }

        // Save user preference
        storeUserPreferences( userPreferences );
    }

    private handleChange = ( name: string, changedStatus: boolean ) => {
        const { userPreferences, onChange } = this.props;
        switch ( name ) {
            case 'pinToMenuBar':
                if ( changedStatus ) {
                    pinLaunchpadToMenu();
                } else {
                    releaseLaunchpadFromMenu();
                }
                break;
            case 'launchOnStart':
                launchOnLogin( changedStatus );
                break;
            default:
                break;
        }

        userPreferences[name] = changedStatus;
        if ( typeof onChange === 'function' ) {
            onChange( userPreferences );
        }
    };

    render() {
        const { userPreferences, requiredItems } = this.props;
        const requiredItemArray = Object.keys( requiredItems );

        return (
            <List aria-label="Preferences">
                {Object.keys( userPreferences ).map( ( userPreference ) => {
                    if ( !requiredItemArray.includes( userPreference ) )
                        return null;

                    return (
                        <PreferenceItem
                            key={generateRandomString()}
                            name={userPreference}
                            status={userPreferences[userPreference]}
                            onChange={this.handleChange}
                            disabled={!requiredItems[userPreference]}
                        />
                    );
                } )}
            </List>
        );
    }
}
