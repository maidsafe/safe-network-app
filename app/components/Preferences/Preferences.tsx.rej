diff a/app/components/Preferences/Preferences.tsx b/app/components/Preferences/Preferences.tsx	(rejected hunks)
@@ -67 +67 @@ export class Preferences extends Component<Props> {
-        if (typeof onChange === 'function') {
+        if ( typeof onChange === 'function' ) {
