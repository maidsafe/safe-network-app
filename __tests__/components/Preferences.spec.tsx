import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';

import { Preferences } from '$Components/Preferences';

const shallow = createShallow();

const shallowSetup = ( propOverrides? ) => {
    const props = Object.assign(
        {
            userPreferences: {
                autoUpdate: false,
                pinToMenuBar: false,
                launchOnStart: false,
                showDeveloperApps: false,
                warnOnAccessingClearnet: false
            },
            onChange: jest.fn()
        },
        propOverrides
    );

    const wrapper = shallow( <Preferences {...props} /> );

    return {
        props,
        wrapper
    };
};

describe( 'Preferences', () => {
    it( 'render', () => {
        const { wrapper } = shallowSetup();
        expect( wrapper ).toMatchSnapshot();
    } );

    it( 'render all preferences by default', () => {
        const { wrapper } = shallowSetup();
        expect( wrapper.children() ).toHaveLength( 2 );
    } );

    it( 'render specific preferences', () => {
        const { wrapper } = shallowSetup( {
            requiredItems: {
                autoUpdate: true,
                pinToMenuBar: true,
                launchOnStart: true
            }
        } );
        expect( wrapper.children() ).toHaveLength( 3 );
        expect(
            wrapper
                .children()
                .at( 0 )
                .props().name
        ).toEqual( 'autoUpdate' );
        expect(
            wrapper
                .children()
                .at( 1 )
                .props().name
        ).toEqual( 'pinToMenuBar' );
        expect(
            wrapper
                .children()
                .at( 2 )
                .props().name
        ).toEqual( 'launchOnStart' );
    } );

    it( 'should disable preference', () => {
        const { wrapper } = shallowSetup( {
            requiredItems: {
                autoUpdate: true,
                pinToMenuBar: false
            }
        } );

        expect( wrapper.children() ).toHaveLength( 2 );
        expect(
            wrapper
                .children()
                .at( 0 )
                .props().disabled
        ).toBeFalsy();
        expect(
            wrapper
                .children()
                .at( 1 )
                .props().disabled
        ).toBeTruthy();
    } );

    it( 'triggers onChange on toggled ', () => {
        const { wrapper, props } = shallowSetup();
        wrapper
            .children()
            .at( 0 )
            .props()
            .onChange();
        expect( props.onChange ).toHaveBeenCalled();
        expect( props.onChange ).toHaveBeenCalledWith( props.userPreferences );
    } );

    it( 'handle no user preferences passed from store', () => {
        const { wrapper } = shallowSetup( {
            userPreferences: {}
        } );
        expect( wrapper.children() ).toHaveLength( 0 );
    } );
} );
