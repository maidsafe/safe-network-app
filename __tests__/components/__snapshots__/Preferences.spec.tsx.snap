// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preferences render 1`] = `ShallowWrapper {}`;
